core/src/main/scala/dagr/core/tasksystem/Pipe.scala

```diff
@@ -0,0 +1,182 @@
+  * before connecting them with pipes.
+  *
+  * See also the traits [[PipeIn]] and [[PipeOut]] which provide a simplified way to implement
+  * a piping task that only receives on or only emits to a pipe.
+  *
+
+  * @tparam In a symbolic type representing the kind of data the task can accept
+  * @tparam Out a symbolic type representing the kind of fdata the task emits
+  */
+trait Pipe[In,Out] extends ProcessTask {
+  /**
+    * Generates or extends a pipe chain by connecting this task to the next task via a pipeline.
+    *
+    * @param next a Pipe enabled task that should receive this pipe's output as its input
+    * @tparam In2 the type of input expected by the next task. Must be the same or a supertype
+    *             of this pipe's output.
+    * @tparam Out2 the type of output of the next task, which becomes the output type of the
+    *              returned pipe chain
+    * @return a new Pipe object that is the result of connecting the two pipes
+    */
+  def |[In2 >: Out,Out2](next: Pipe[In2,Out2]): Pipe[In,Out2] = {
+    Pipes.chain(this, next)
+  }
+
+  /** Returns the first task in the pipe. */
+  private[tasksystem] def left: Pipe[In,_] = this
+  /** Returns the last task in the pipe. */
+  private[tasksystem] def right: Pipe[_,Out] = this
+  /** Returns the ordered set of Pipes/tasks in the chain from first to last inclusive. */
+  private[tasksystem] def ordered: Seq[Pipe[_,_]] = List[Pipe[_,_]](this)
+}
+
+/** A simplified trait for sink tasks that can receive data via a pipe, but cannot pipe onwards. */
+trait PipeIn[In] extends Pipe[In,Nothing]
+
+/** A simplified trait for generate tasks that can pipe out, but cannot recieve data from a pipe. */
+trait PipeOut[Out] extends Pipe[Nothing,Out]
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Section: Objects and private utility classes to make it all work
+///////////////////////////////////////////////////////////////////////////////
+object Pipes {
+  /**
+    * A class that represents an empty pipe, that allows for easier construction of pipes with conditional branches.
+    * Must remain private and only be intantiated by Piping.empty().
+    */
+  private class EmptyPipe[T] extends Pipe[T,T] {
+    override def args: Seq[Any] = throw new IllegalStateException("Should not reach here.")
+    override def pickResources(availableResources: ResourceSet): Option[ResourceSet] = throw new IllegalStateException("Should not reach here.")
+  }
+
+  /**
+    * An immutable class representing a chain of tasks that can be piped together. Currently only supports chains
+    * with up to a maximum of one non-FixedResources task, in order to keep the resource management code simple.
+    *
+    * The constructor does not check types as thoroughly as one might like, instead this is done in
+    * [[Pipes.chain()]], which is the only place the constructor is called from.
```

Comment:
So, I'm predisposed against doing this for a couple of reasons:
1. My comment here was not that things are un-typesafe, just that the typesafety is enforced elsewhere, so we need to be a touch more careful editing this code not to break type safety by making this class or constructor non-private
2. Requiring `ClassTag`s or `TypeTags` makes life more difficult for users.  It's infectious: any code that uses type parameters and your API, also has to require `TypeTag`s, so they'll be available when your API is called.  E.g. we'd have to rewrite `FifoBuffer` to also take a TypeTag since it's generic and doesn't care about it's data type.  Then in client code you couldn't write:
```scala
def buffer[T] = new FifoBuffer().requires(memory="1G").withName("buffer" + counter++)
```
because you'd get some esoteric compiler warnings about TypeTags, and then you'd have to write:
```scala
def buffer[T: TypeTag] = new FifoBuffer().requires(memory="1G").withName("buffer" + counter++)
```
3. The types are all checked by the compiler as it stands (try it!). I don't think runtime enforcement is necessary.  If you cast things and do bad things, that's you're problem.  E.g. try this in the REPL:
```scala
import scala.collection.mutable.ListBuffer
val xs = ListBuffer[String]()
val ys = xs.asInstanceOf[ListBuffer[Any]]
ys.append(1)
xs foreach ((x: String) => println(x))
```
It compiles just find and gives a cast exception at runtime.